File: example/java/jimmer-sql/model/src/main/java/org/babyfish/jimmer/sql/example/model/TreeNode.java
package org.babyfish.jimmer.sql.example.model;

import org.babyfish.jimmer.sql.Key;

import org.babyfish.jimmer.sql.*;
import org.babyfish.jimmer.sql.example.model.common.BaseEntity;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@Entity
public interface TreeNode extends BaseEntity {

    @Id
    @Column(name = "NODE_ID")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key // ❶
    String name();

    @Nullable // ❷ Null property, Java API requires this annotation, but kotlin API does not
    @Key // ❸
    @ManyToOne // ❹
    @OnDissociate(DissociateAction.DELETE) // ❺
    TreeNode parent();

    @OneToMany(mappedBy = "parent", orderedProps = @OrderedProp("name")) // ❻
    List<TreeNode> childNodes();
}

/*----------------Documentation Links----------------
❶ ❸ https://babyfish-ct.github.io/jimmer/docs/mapping/advanced/key
❷ https://babyfish-ct.github.io/jimmer/docs/mapping/base/nullity
❹ https://babyfish-ct.github.io/jimmer/docs/mapping/base/association/many-to-one

❺ https://babyfish-ct.github.io/jimmer/docs/mapping/advanced/on-dissociate
  https://babyfish-ct.github.io/jimmer/docs/mutation/save-command/dissociation
  https://babyfish-ct.github.io/jimmer/docs/mutation/delete-command

❻ https://babyfish-ct.github.io/jimmer/docs/mapping/base/association/one-to-many
---------------------------------------------------*/

File: example/kotlin/jimmer-sql-kt/model/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/model/TreeNode.kt
package org.babyfish.jimmer.sql.example.kt.model

import org.babyfish.jimmer.sql.*
import org.babyfish.jimmer.sql.example.kt.model.common.BaseEntity

@Entity
interface TreeNode : BaseEntity {

    @Id
    @Column(name = "NODE_ID")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    @Key // ❶
    val name: String

    @Key // ❷
    @ManyToOne // ❸
    @OnDissociate(DissociateAction.DELETE) // ❹
    val parent: TreeNode?

    @OneToMany(mappedBy = "parent", orderedProps = [OrderedProp("name")]) // ❺
    val childNodes: List<TreeNode>
}

/*----------------Documentation Links----------------
❶ ❷ https://babyfish-ct.github.io/jimmer/docs/mapping/advanced/key
❸ https://babyfish-ct.github.io/jimmer/docs/mapping/base/association/many-to-one

❹ https://babyfish-ct.github.io/jimmer/docs/mapping/advanced/on-dissociate
  https://babyfish-ct.github.io/jimmer/docs/mutation/save-command/dissociation
  https://babyfish-ct.github.io/jimmer/docs/mutation/delete-command

❺ https://babyfish-ct.github.io/jimmer/docs/mapping/base/association/one-to-many
---------------------------------------------------*/